### PR TITLE
Azure Rightsize SQL Instances

### PR DESCRIPTION
**Pull request by srpomeroy**
### Description

Identifies underutilized Azure Databases

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/azure/azure-sql-utilization/azure_databases_utilization.pt [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+###############################################################################
+datasource "ds_azure_databases" do
+  request do
+    auth $azure_auth
+    pagination $azure_pagination
+    host "management.azure.com"
+    path join(["/subscriptions/", $param_azure_sub, "/resources"])
+
+    query "api-version","2019-08-01"
+#    query "$filter", "resourceType eq 'Microsoft.Sql/servers/databases' or resourceType eq 'Microsoft.DocumentDb/databaseAccounts'"
```
Comment (rshade) [inline cost/azure/azure-sql-utilization/azure_databases_utilization.pt]: Can we remove this?

File: cost/azure/azure-sql-utilization/azure_databases_utilization.pt [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        "User-Agent" : "RS Policies"
+      }
+    } 
+  EOS
+end
+
+datasource "ds_azure_databases_metrics" do
+  iterate $ds_azure_databases
+  request do
+    run_script $utilization_request, val(iter_item,"id")
```
Comment (rshade) [inline cost/azure/azure-sql-utilization/azure_databases_utilization.pt]: this should be js_

File: cost/azure/azure-sql-utilization/azure_databases_utilization.pt [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+      }
+    }
+  EOS
+end
+
+###############################################################################
+# Policy
+###############################################################################
+
+policy 'policy_azure_db_uttialization' do
```
Comment (rshade) [inline cost/azure/azure-sql-utilization/azure_databases_utilization.pt]: Utilization

File: cost/azure/azure-sql-utilization/azure_databases_utilization.pt [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+  validate_each $ds_merged_metrics do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{ len data }} Underutilized Instances Reporting to Azure Log Analytics"
+    detail_template <<-EOS
+# Azure Dababase Utilization
```
Comment (rshade) [inline cost/azure/azure-sql-utilization/azure_databases_utilization.pt]: Databsase

File: cost/azure/azure-sql-utilization/azure_databases_utilization.pt [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+| ------ | ---- | ---- | ---- | -----------| ------------ | ---------- | -------------- | ------------- |
+{{ range data -}}
+| {{.location}} | {{.name}} | {{.kind}} | {{.type}} | {{.sku.name}} | {{.sku.family}} | {{.sku.tier}} | {{.sku.tier}} | {{ .average_cpu }} |
+{{ end -}}
+
+___
+###### Policy Applied in Account: {{ rs_project_name }} (Account ID: {{ rs_project_id }}) within Org: {{ rs_org_name }} (Org ID: {{ rs_org_id }})
+EOS
+    escalate $email
+##    check logic_or(eq($param_avg_cpu, -1), gt(to_n(val(item,"cpu_average")),$param_avg_cpu))
```
Comment (rshade) [inline cost/azure/azure-sql-utilization/azure_databases_utilization.pt]: Need the actual logic back.

File: cost/azure/rightsize_sql_instances/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+- *Exclusion Tag Key* - To Identify any instances that should be excluded from the recommendations
+- *Email addresses of the recipients you wish to notify* - A list of email addresses to notify
+
+### Required RightScale Roles
+
+- `credential_viewer`
+
+### Azure Required Permissions
+
+- `Read: Get SQL Server registration details`
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/README.md]: Lets add the permissions needed here just like the other policies

File: cost/azure/rightsize_sql_instances/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# Azure Rightsize SQL Instances
+
+## What it does
+
+This policy will look at Utilization of Azure SQL instances and recommend up or down sizing.
+
+## Functional Details
+
+This policy checks all the Azure SQL instances for a Azure Subscription. It does a Average CPU usage over the last 30 days. It then checks if the Utilization is Lower than the Downsize Threshold or higher that UpScale Threshold. Finally it display the found data and the recommendations.
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/README.md]: use upsize instead of upscale.

File: cost/azure/rightsize_sql_instances/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+### Input Parameters
+
+- *Azure Tenant ID* - Your Azure tenant ID.  You can find it by following this guide: [Tenant ID](https://docs.microsoft.com/en-us/onedrive/find-your-office-365-tenant-id)
+- *Azure Subscription ID* - Your Azure Subscription ID.  You can find it by following this guide: [Subscription ID](https://blogs.msdn.microsoft.com/mschray/2016/03/18/getting-your-azure-subscription-guid-new-portal/)
+- *Average used CPU % - Upscale threshold* - Percentage of CPU utilization to identify an Upscale is recommended
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/README.md]: use upsize

File: cost/azure/rightsize_sql_instances/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+- *Average used CPU % - Downscale Threshold* - Percentage of CPU utilization to identify an Downscale is recommended
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/README.md]: use downsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+parameter "param_azure_sub" do
+  type "string"
+  label "Azure Subscription ID"
+  category "Azure"
+end
+
+parameter "param_avg_cpu_upscale" do
+  type "number"
+  label "Average used CPU % - Upscale threshold"
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: use upsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+  description "Set to -1 to ignore CPU utilization"
+  default 80
+  min_value -1
+  max_value 100
+end
+
+parameter "param_avg_cpu_downscale" do
+  type "number"
+  label "Average used CPU % - Downscale Threshold"
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: use downsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+        result[i].average_cpu = average_cpu.toFixed(2);
+      }  else {
+        result[i].average_cpu = "n/a";
+      }
+
+      //Now check for Recomendation
+      if (result[i].average_cpu !== "n/a" && ((Number(result[i].average_cpu) >= Number(up_threshold)) || (Number(result[i].average_cpu) <= Number(down_threshold)))) {
+        if (Number(result[i].average_cpu) >= Number(up_threshold)) {
+          //Upscale
+          result[i].recomendation = "Scale up";
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: Use Upsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+        } else {
+          //Downscale
+          result[i].recomendation = "Scale Down";
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: use Downsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+  validate_each $ds_merged_metrics do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{ len data }} Azure Rightsize SQL Instances"
+    detail_template <<-EOS
+# Azure Rightsize SQL Instances
+
+### Thresholds for Consideration
+  - Upscale Average CPU% threshold   : {{ parameters.param_avg_cpu_upscale }} 
+  - Downscale Average CPU% threshold : {{ parameters.param_avg_cpu_downscale }}
+
+| Region | Name | Resource Group | Kind | Type | SKU - Name | SKU - family | SKU - Tier | SKU - Capacity | CPU Average % | recomendation |
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: Capitalize Recommendation, and check spelling.  Two m's

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+###############################################################################
+
+policy 'policy_azure_db_utilization' do
+  validate_each $ds_merged_metrics do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{ len data }} Azure Rightsize SQL Instances"
+    detail_template <<-EOS
+# Azure Rightsize SQL Instances
+
+### Thresholds for Consideration
+  - Upscale Average CPU% threshold   : {{ parameters.param_avg_cpu_upscale }}
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: use Upsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+
+policy 'policy_azure_db_utilization' do
+  validate_each $ds_merged_metrics do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{ len data }} Azure Rightsize SQL Instances"
+    detail_template <<-EOS
+# Azure Rightsize SQL Instances
+
+### Thresholds for Consideration
+  - Upscale Average CPU% threshold   : {{ parameters.param_avg_cpu_upscale }} 
+  - Downscale Average CPU% threshold : {{ parameters.param_avg_cpu_downscale }}
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: use Downsize

File: cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt [hunk 17]
```diff
@@ -0,0 +1,273 @@
+name "Azure Rightsize SQL Instances"
+rs_pt_ver 20180301
+type "policy"
+short_description "Gathers Azure SQL database utilization data from Azure Metrics, and reports on databases that is recommended to be resized. See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/azure/rightsize_sql_instances/) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
```
Comment (cdwilhelm) [inline cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: use: Finds Azure SQL instances based on CPU and Memory thresholds and Resize them after approval.

Comment (kim-ras) [reply cost/azure/rightsize_sql_instances/azure_rightsize_sql_instances.pt]: The requirement is only recommendation.  Can we agree to push the resize to a future version ?

**Pull request merged**